README.md

### ckan-install

Installing git, Docker, Docker Compose, CKAN, pdf_view ckan extension

## PREREQUISITES:
- Uncomment globals for (staging or development)
- Root access

## INSTRUCTIONS:
- Run script from root

# ------------------------------------------------------------

### ckan-deploy

Installing rsync, python-pip, ckanapi.
Transfering resources and users from one CKAN instance to another.
Dumping and loading organizations, datasets, groups from one CKAN instance to another.

## PREREQUISITES:
- Must have two scripts: 	
	ckan-install (installation of new CKAN)
	ckan-deploy (migration from existing CKAN instance (SRC_CKAN) to a new one (DEST_CKAN))
- (Except for fromProdToStaging option) Must have direct ssh to SRC_CKAN (which must be running)
- User at DEST_CKAN must have sudo access

## INSTRUCTIONS:
- First, open ckan-install script, read instructions and run it
- In your browser, open installed CKAN, log in as sysadmin, go to sysadmins profile, 
		find api-key and put it in api_ckan_sysadmin_ (in this script)
- Then, in this script, change other variables (if needed), addr_user_, addr_port_, etc., 
- Run this script
- Optional: open .env file (/path/to/ckan/contrib/docker/.env),
			change password for database ('ckan' by default), and restart containers
- Configure website manually (name, about, logo, etc.)


CHANGELOG.md
### ckan-install

# [1.1] 2019-07-22
## Added: 
- Globals for production

# [1.2] 2019-07-25
## Added: 
- Global variables for "Staging on P17SM at Ecole de la paix"
- Creation of $myPath directory, if doesn't exist

## Updated:
- Install Docker-Compose 1.24.1 instead of 1.24.0

## Removed:
- TODO comment


# --------------------------------------------------------------------------------

### ckan-deploy

# [2.0] 2019-07-22
## Added:
- Script for fromStagingToProd conditional option
- Script for fromProdToStaging conditional option
- Globals (variables) for production
- Globals (variables) for staging
- Instructions for changing database password, for secutiry reasons
- Instructions for website configuration
- Function install_ckan_migration_tools() to avoid repetition
- Function ckanapi_dump_load() to avoid repetition

## Updated:
- Using install_ckan_migration_tools() function instead of repeating code in every conditional option
- Using ckanapi_dump_load() function instead of repeating code in every conditional option

# [2.1] 2019-07-23
## Added:
- CHANGELOG.md file

## Updated:
- Fixed rsync for # Transfer resource files
- More readable instruction

# [3.0] 2019-07-23
## Added:
- Function deploy(): abstraction to avoid repetition
- Call to deploy() at fromStagingToDev and fromStagingToProd

## Removed:
- Repetetive script at fromStagingToDev and fromStagingToProd
- echo at install_ckan_migration_tools() function

# [3.1] 2019-07-24
## Updated:
- Fixed (cleaned) fromProdToStaging script

## Removed:
- `--upload-resources`  option at ckanapi: caused server error, because couldn't access SRC_CKAN url: localhost:5000


# --------------------------------------------------------------------------------

### production.ini
## Added:
- licenses_group_url = `file:///var/lib/docker/volumes/docker_ckan_home/_data/venv/src/ckan.json`
	reference: https://boykoc.github.io/ckan/configuration/2018/06/13/adding-custom-licenses-in-ckan.html


# --------------------------------------------------------------------------------

### ckan.json
## Added:
- CC-BY-3.0-IGO
- CC-BY-NC-SA-4.0
- other

## Updated:
- moved to `/var/lib/docker/volumes/docker_ckan_home/_data/venv/src/`
- chown 900:900 `/var/lib/docker/volumes/docker_ckan_home/_data/venv/src/ckan.json`
- chmod 644 `/var/lib/docker/volumes/docker_ckan_home/_data/venv/src/ckan.json `


ckan-install.sh
#!/bin/bash
# ckan-install_2019-07-25_v1.2
# See CHANGELOG.md and README.md for prerequisites and instructions
# contact: nurtay@protonmail.ch


## GLOBALS 

# For staging:
#deb_ver='buster'
#usr="username"
#CKAN_URL='example.fr'
#CKAN_PORT='5000'

# Staging on P17SM at Ecole de la paix
deb_ver='disco' # Ubuntu 19.04
usr="username"
CKAN_URL='localhost'
CKAN_PORT='5000'

# For development:
#deb_ver='stretch'
#usr="osboxes"
#CKAN_URL='localhost'
#CKAN_PORT='5000'

# For production
# deb_ver='stretch'
# usr="username"
# CKAN_URL='example.fr'
# CKAN_PORT='5000'

# For all:
myPath="/home/$usr/bin/ckan" # directory where CKAN to be cloned
# VOL_CKAN_CONFIG="/var/lib/docker/volumes/docker_ckan_config/_data/"
VOL_CKAN_CONFIG="/var/snap/docker/common/var-lib-docker/volumes/docker_ckan_config/_data/"
# tmp='/etc/apt/sources.list'
mkdir -p $myPath 

# Exit if any error occurs
set -e

# Echo after each installation
fin() {
  echo "------> Finished: $1"
}

# ************************************************************* Install git
read -p "Do you want to install git? (y | n) " -r
echo    
if [[ $REPLY =~ ^[Yy]$ ]]; then
    # Comment deb cdrom
    sed -i '/^deb cdrom/s/^/#/' $tmp
    fin "Comment deb cdrom"

	# Add one line in sources.list and update
	echo "deb http://deb.debian.org/debian/ $deb_ver main contrib" >> $tmp
	apt-get update 
	yes | apt-get upgrade 
	yes | apt-get dist-upgrade
	yes | apt-get -f install
	fin "Add one line in sources.list and update"

	apt update
	yes | apt install git
	git --version
	fin "Install git"
fi

# ************************************************************* Install Docker
read -p "Do you want to install Docker? (y | n) " -r
echo    
if [[ $REPLY =~ ^[Yy]$ ]]; then
	# Install packages to allow apt to use a repository over HTTPS
	yes | apt-get install \
	    apt-transport-https \
	    ca-certificates \
	    curl \
	    gnupg2 \
	    software-properties-common
	fin "Install packages to allow apt to use a repository over HTTPS"

	# Add Docker’s official GPG key
	curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add -
	fin "Add Docker’s official GPG key"

	# Use the following command to set up the stable repository. To add the nightly or test repository, add the word nightly or test (or both) after the word stable
	add-apt-repository \
	 "deb [arch=amd64] https://download.docker.com/linux/debian \
	 $(lsb_release -cs) \
	 stable"
	fin "set up the stable repository"

	# Install the latest version of Docker CE and containerd, or go to the next step to install a specific version
	apt-get update
	yes | apt-get install docker-ce docker-ce-cli containerd.io
	fin "Install the latest version of Docker CE and containerd, or go to the next step to install a specific version"

	# Verify that Docker CE is installed correctly by running the hello-world image
	docker run hello-world
	fin "Verify that Docker CE is installed correctly by running the hello-world image"

	# Add your user to the docker group.
	usermod -aG docker $usr # $USER
	fin "Add your user to the docker group"
fi

# ************************************************************* Install Docker Compose
read -p "Do you want to install Docker Compose? (y | n) " -r
echo    
if [[ $REPLY =~ ^[Yy]$ ]]; then
    # Download the current stable release of Docker Compose
	curl -L "https://github.com/docker/compose/releases/download/1.24.1/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
	fin "Download the current stable release of Docker Compose"

	# Apply executable permissions to the binary
	chmod +x /usr/local/bin/docker-compose
	fin "Apply executable permissions to the binary"

	docker-compose --version
	fin "Install Docker Compose"
fi

# ************************************************************* Install CKAN
# Clone stable CKAN into a directory of your choice
cd $myPath
git clone --branch ckan-2.8.2 https://github.com/ckan/ckan.git
fin "Clone CKAN"

# Sensitive settings and environment variables, change URL and PORT, add CKAN_MAX_UPLOAD_SIZE_MB
cd $myPath/ckan/contrib/docker
cp .env.template .env
sed -i s/localhost:5000/$CKAN_URL:$CKAN_PORT/g .env
sed -i s/CKAN_PORT=5000/CKAN_PORT=$CKAN_PORT/g .env
echo "CKAN_MAX_UPLOAD_SIZE_MB=500" >> .env
fin "Sensitive settings and environment variables, change URL and PORT, add CKAN_MAX_UPLOAD_SIZE_MB"

# Build the images
cd $myPath/ckan/contrib/docker
docker-compose up -d --build
fin "Build the images"

# If the CKAN logs show problems connecting to the database, restart the ckan container a few times
cd $myPath/ckan/contrib/docker
docker-compose restart ckan
docker-compose restart ckan
fin "If the CKAN logs show problems connecting to the database, restart the ckan container a few times"

# Configure datastore database
docker exec ckan /usr/local/bin/ckan-paster --plugin=ckan datastore set-permissions -c /etc/ckan/production.ini | docker exec -i db psql -U ckan
fin "Configure datastore database"

# Add datastore datapusher to ckan.plugins and enable the datapusher option ckan.datapusher.formats
sed -i '/^ckan.plugins/s/$/ datastore datapusher/' $VOL_CKAN_CONFIG/production.ini
sed -i '/^#.*ckan.datapusher.formats/s/^#//' $VOL_CKAN_CONFIG/production.ini
fin "Add datastore datapusher to ckan.plugins and enable the datapusher option ckan.datapusher.formats"

# Restart the ckan container to apply changes to the production.ini
# cd $myPath/ckan/contrib/docker
# docker-compose restart ckan
# fin "Restart the ckan container to apply changes to the production.ini"

# Rename containers (ex: db -> ckan_db)
cd $myPath/ckan/contrib/docker
sed -Ei 's/(container_name:\s)datapusher/\1ckan_datapusher/' docker-compose.yml
sed -Ei 's/(container_name:\s)db/\1ckan_db/' docker-compose.yml
sed -Ei 's/(container_name:\s)solr/\1ckan_solr/' docker-compose.yml
sed -Ei 's/(container_name:\s)redis/\1ckan_redis/' docker-compose.yml
fin "Rename containers (ex: db -> ckan_db)"

# Create the CKAN admin user (admin is the login)
# email: admin@example.org
# pass: JohnDoe1
docker exec -it ckan /usr/local/bin/ckan-paster --plugin=ckan sysadmin -c /etc/ckan/production.ini add admin
fin "Create the CKAN admin user (admin is the login)"

# -------------- Add pdf_view extension
# Inside the running container, activate the virtualenv and Install the ckanext-pdfview
docker exec -it ckan /bin/bash -c 'source $CKAN_VENV/bin/activate && cd $CKAN_VENV/src/ && pip install ckanext-pdfview'
fin "Inside the running container, activate the virtualenv and Install the ckanext-pdfview"
# Add pdf_view to the ckan.plugins setting 
#                     ckan.views.default_views
sed -i '/^ckan.plugins/s/$/ pdf_view/' $VOL_CKAN_CONFIG/production.ini
sed -i '/^ckan.views.default_views/s/$/ pdf_view/' $VOL_CKAN_CONFIG/production.ini
fin "Add pdf_view to the ckan.plugins setting and to ckan.views.default_views"
cd $myPath/ckan/contrib/docker
docker-compose stop && docker-compose up -d
fin "Add pdf_view extension"


:<<'MIGHT_NEED'
# Fix SSL problem
yes | apt install python-pip
pip --version
pip install requests[security]
fin "Fix SSL problem"

# max upload size
cd $myPath/ckan/ckanext/datapusher
docker cp datapusher:/usr/src/app/datapusher/jobs.py .
cd $myPath/ckan/contrib/docker
echo "CKAN_MAX_UPLOAD_SIZE_MB=15" >> .env
fin "max upload size"
MIGHT_NEED


ckan-deploy.sh
#!/bin/bash
# ckan-deploy_2019-07-23_v3.1
# See CHANGELOG.md and README.md for prerequisites and instructions
# contact: nurtay@protonmail.ch

# Exit if any error occurs
set -e

# Globals
path_ckan_vol_storage='/var/lib/docker/volumes/docker_ckan_storage/_data'
path_ckan_vol_home='/var/lib/docker/volumes/docker_ckan_home/_data'
# Staging
path_tmp_staging='/home/username/Documents' # temporary path at staging
path_ckan_docker_compose_staging='/home/username/ckan/contrib/docker'
addr_user_staging='username@example.fr'
addr_port_staging='http://example.fr:8082'
api_ckan_sysadmin_staging=''
# Development
path_ckan_docker_compose_dev='/home/osboxes/ckan/contrib/docker'
addr_port_dev='http://localhost:5000'
api_ckan_sysadmin_dev=''
# Production
path_tmp_prod='/home/osboxes/Documents' # temporary path at prod sever
path_ckan_docker_compose_prod='/home/osboxes/ckan/contrib/docker'
addr_port_prod='http://localhost:5000'
api_ckan_sysadmin_prod=''

function help() {
	printf "Usage:  $0 [fromStagingToDev | fromStagingToProd | fromDevToStaging | fromProdToStaging]\n"
	return 1
}

# Notifies if a command was executed successfully
function ok() {
	printf "||||||| FINISHED: $1\n"
}

# Migration requires rsync, python-pip, ckanapi
# The following function proposes to install them
function install_ckan_migration_tools() {
	read -p "Do you want to install rsync? (y | n) " -r 
	if [[ $REPLY =~ ^[Yy]$ ]]; then
		yes | sudo apt-get install rsync
		ok "Install rsync"
	fi

	read -p "Do you want to install python-pip? (y | n) " -r
	if [[ $REPLY =~ ^[Yy]$ ]]; then
		sudo apt update
		yes | sudo apt install python-pip
		ok "Install pip"
	fi

	read -p "Do you want to install ckanapi? (y | n) " -r
	if [[ $REPLY =~ ^[Yy]$ ]]; then
		sudo pip install ckanapi
		ok "Install ckanapi"
	fi
	ok "Install CKAN migration tools"
}

# Function exports and uploads groups, orgs, datasets
# Usage: ckanapi_dump_load $addr_port_prod $addr_port_staging $api_ckan_sysadmin_staging
function ckanapi_dump_load() {
	# Note that $1 ($addr_port_[staging|dev|prod]) must be running
	# Export data
	sudo ckanapi dump organizations --all -O orgs.jsonl.gz -z -r $1
	sudo ckanapi dump datasets --all -O dsets.jsonl.gz -z -r $1
	sudo ckanapi dump groups --all -O groups.jsonl.gz -z -r $1
	ok "Export data"
	# Upload data
	sudo ckanapi load organizations -I orgs.jsonl.gz -z -r $2 -a $3
	sudo ckanapi load datasets -I dsets.jsonl.gz -z -r $2 -a $3 
	# previous version was: sudo ckanapi load datasets --upload-resources -I dsets.jsonl.gz -z -r $2 -a $3
	# but --upload-resources option didn't work, because of SRC_CKAN url: localhost:5000
	sudo ckanapi load groups --upload-logo -I groups.jsonl.gz -z -r $2 -a $3
	ok "Upload data"
}

# Function should take 3 parameters: 
# $path_ckan_docker_compose_[dev|prod]
# $addr_port_[dev|prod]
# $api_ckan_sysadmin_[dev|prod]
function deploy() {
	# Install CKAN migration tools
	install_ckan_migration_tools

	# At staging server: copy _data/ && run SRC_CKAN && user.sql available at staging
	printf "\n*****************************************************************************************\n"
	printf "IMPORTANT: After providing password for ssh, run the following commands one by one (look script for more):\n"
	# We have to copy, because otherwise we get Permission denied (13), because
	# rsync needs sudo access on the other side of ssh pipe
	printf "sudo cp -r $path_ckan_vol_storage $path_tmp_staging\n"
	# Export the user table
	printf "cd $path_ckan_docker_compose_staging && docker-compose up -d\n"
	printf "docker exec -it ckan_db /bin/bash -c 'pg_dump -h localhost -U ckan -a -O -t user -f user.sql ckan && exit'\n"
	printf "docker cp ckan_db:/user.sql $path_tmp_staging\n"
	printf "exit\n"
	printf "\n*****************************************************************************************\n"
	ssh -p 22 $addr_user_staging
	ok "At staging: copy _data/ && run SRC_CKAN && user.sql available at staging"

	# Transfer resource files
	sudo rsync -Pavvre "ssh -p 22" $addr_user_staging:$path_tmp_staging/_data $path_ckan_vol_storage
	ok "Transfer resource files"

	# Transfer users
	# Make user.sql accessible to the source CKAN container
	# and chown user.sql to the owner of $CKAN_HOME (ckan, UID 900)
	sudo rsync -Pavvre "ssh -p 22" $addr_user_staging:$path_tmp_staging/user.sql $path_ckan_vol_home/venv/src
	sudo chown 900:900 $path_ckan_vol_home/venv/src/user.sql
	ok "user.sql accessible to SRC_CKAN container && chown user.sql"
	# Now the file user.sql is accessible from within the ckan container
	cd $1 && docker-compose up -d
	printf "Password for user ckan is at $1/.env\n"
	docker exec -it ckan /bin/bash -c 'psql -U ckan -h db -f $CKAN_VENV/src/user.sql'
	ok "Transfer users"

	# Export and upload groups, orgs, datasets
	# Note that $addr_port_staging must be running
	ckanapi_dump_load $addr_port_staging $2 $3

	# Rebuild search index. Trigger a Solr index rebuild:
	docker exec -it ckan /usr/local/bin/ckan-paster --plugin=ckan search-index rebuild -c /etc/ckan/production.ini
	ok "Trigger a Solr index rebuild"

	# Cleaning up
	printf "\n*****************************************************************************************\n"
	printf "IMPORTANT: After providing password for ssh, run the following commands one by one (look script for more):\n"
	printf "cd $path_tmp_staging && sudo rm -rfv _data user.sql\n"
	printf "exit\n"
	printf "\n*****************************************************************************************\n"
	ssh -p 22 $addr_user_staging
	ok "At staging: remove _data, user.sql"
	sudo rm -v orgs.jsonl.gz dsets.jsonl.gz groups.jsonl.gz
	ok "Remove orgs.jsonl.gz dsets.jsonl.gz groups.jsonl.gz"
	cd $1 && docker-compose stop && docker-compose up -d
	ok "Deployment"
}

# Main program
if [[ $# -ne 1 || $1 = "--help" || $1 = "-h" ]]; then help

elif [[ $1 = "fromStagingToDev" ]]; then
	deploy $path_ckan_docker_compose_dev $addr_port_dev $api_ckan_sysadmin_dev

elif [[ $1 = "fromStagingToProd" ]]; then
	deploy $path_ckan_docker_compose_prod $addr_port_prod $api_ckan_sysadmin_prod

elif [[ $1 = "fromDevToStaging" ]]; then
	printf "TODO\n"

elif [[ $1 = "fromProdToStaging" ]]; then
	# TODO
else
	help
fi